lib/typecheck.ts

```diff
@@ -679,6 +679,61 @@ export default class TypeChecker {
         return Promise.resolve();
     }
 
+    private async _typeCheckJoin(ast : Ast.JoinExpression, scope : Scope) {
+        assert(ast.lhs.schema && ast.rhs.schema);
+        const name = `join(${ast.lhs.schema.name},${ast.rhs.schema.name})`;
+        if (!ast.lhs.schema.is_list || !ast.rhs.schema.is_list)
+            throw new TypeError('Both sides of a join must be a list query');
+        const classDef = null; // FIXME: what should this be for a join expression? 
+        const qualifiers = {
+            is_list : true,
+            is_monitorable: ast.lhs.schema.is_monitorable || ast.rhs.schema.is_monitorable
+        };
+        const args = [];
+        for (const arg of ast.lhs.schema.iterateArguments()) {
+            if (arg.is_input)
+                continue;
+            const newArg = arg.clone();
+            newArg.name = `first.${arg.name}`;
+            args.push(newArg);
+        }
+        for (const arg of ast.rhs.schema.iterateArguments()) {
+            if (arg.is_input)
+                continue;
+            const newArg = arg.clone();
+            newArg.name = `second.${arg.name}`;
+            args.push(newArg);
+        }
+        ast.schema = new Ast.FunctionDef(null, 'query', classDef, name, [], qualifiers, args);
+        if (ast.condition) 
+            await this._typeCheckJoinCondition(ast.condition, ast.schema);
+        return Promise.resolve();
+    }
+
+    private async _typeCheckJoinCondition(ast : Ast.BooleanExpression, schema : Ast.FunctionDef) {
+        if (ast instanceof Ast.AtomBooleanExpression) {
+            if (!ast.name.startsWith('first.'))
+                throw new TypeError('Fields on the left-hand side in a join condition needs a `first.` prefix');
```

Comment:
Why do you need the `first.` prefix? I thought the point of the special join syntax was to avoid `first.` and `second.` in the common cases by relying on the operand order...

Reply:
I mentioned this in the commit message. The observation from examples is that it's actually less readable and more confusing without them. Another minor issue is that it's a bit weird that the fields do not match the schema when compiling the program.

Reply:
But then, what's the difference between:
```
@q1() join @q2() on first.foo = second.bar
```
and
```
(@q1() join @q2()) filter first.foo = second.bar
```
?

The special join syntax is unnecessary.

Reply:
good point... I think at this point I should just drop that special syntax